util/logtools/encounter_tools.ts

```diff
@@ -0,0 +1,283 @@
+import { UnreachableCode } from '../../resources/not_reached';
+import ZoneInfo from '../../resources/zone_info';
+import { NetMatches, NetAnyMatches } from '../../types/net_matches';
+import { CactbotBaseRegExp } from '../../types/net_trigger';
+import { commonReplacement, syncKeys } from '../../ui/raidboss/common_replacement';
+
+// TODO: add some error checking that a zone has been found before a fight.
+// This can happen on partial logs.
+
+type ZoneEncInfo = {
+  name?: string;
+  zoneId?: number;
+  startLine?: string;
+  endLine?: string;
+  startTime?: Date;
+  endTime?: Date;
+};
+
+type FightEncInfo = {
+  name?: string;
+  zoneId?: number;
+  startLine?: string;
+  endLine?: string;
+  startTime?: Date;
+  endTime?: Date;
+  endType?: string;
+  sealName?: string;
+};
+
+export class EncounterFinder {
+  currentZone: ZoneEncInfo = {};
+  currentFight: FightEncInfo = {};
+  currentSeal?: string;
+  zoneInfo?: typeof ZoneInfo[number];
+
+  haveWon = false;
+  haveSeenSeals = false;
+
+  regex: {
+    changeZone: CactbotBaseRegExp<'ChangeZone'>;
+    cactbotWipe: CactbotBaseRegExp<'GameLog'>;
+    win: CactbotBaseRegExp<'ActorControl'>;
+    wipe: CactbotBaseRegExp<'ActorControl'>;
+    commence: CactbotBaseRegExp<'ActorControl'>;
+    playerAttackingMob: CactbotBaseRegExp<'Ability'>;
+    mobAttackingPlayer: CactbotBaseRegExp<'Ability'>;
+  };
+
+  sealRegexes: Array<CactbotBaseRegExp<'GameLog'>> = [];
+  unsealRegexes: Array<CactbotBaseRegExp<'GameLog'>> = [];
+
+  initializeZone(): void {
+    this.currentZone = {};
+  }
+  initializeFight(): void {
+    this.currentFight = {};
+  }
+  constructor() {
+    this.regex = {
+      changeZone: NetRegexes.changeZone(),
+      cactbotWipe: NetRegexes.echo({ line: 'cactbot wipe.*?' }),
+      win: NetRegexes.network6d({ command: '40000003' }),
+      wipe: NetRegexes.network6d({ command: '40000010' }),
+      commence: NetRegexes.network6d({ command: '4000000[16]' }),
+      playerAttackingMob: NetRegexes.ability({ sourceId: '1.{7}', targetId: '4.{7}' }),
+      mobAttackingPlayer: NetRegexes.ability({ sourceId: '4.{7}', targetId: '1.{7}' }),
+    };
+
+    const sealReplace = commonReplacement.replaceSync[syncKeys.seal];
+    if (!sealReplace)
+      throw new Error('missing seal regex');
+    for (const regexString of Object.values(sealReplace)) {
+      const line = `${regexString?.replace('$1', '(?<seal>.*?)')}.*?`;
+      this.sealRegexes.push(NetRegexes.message({ line: line }));
+    }
+
+    const unsealReplace = commonReplacement.replaceSync[syncKeys.unseal];
+    if (!unsealReplace)
+      throw new Error('missing unseal regex');
+    for (const lang of Object.values(unsealReplace)) {
+      const line = `*.?${lang}.*?`;
+      this.unsealRegexes.push(NetRegexes.message({ line: line }));
+    }
+  }
+
+  skipZone(): boolean {
+    if (!this.currentZone.zoneId || !this.zoneInfo)
+      return false;
+    if (!(this.zoneInfo.contentType))
```

Comment:
tiny nit: unnecessary parens

```suggestion
    if (!this.zoneInfo.contentType)
```